pkg/kubernetes/deployment/extnv1beta1/v1alpha1/deployment.go

```diff
@@ -0,0 +1,214 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	extnv1beta1 "k8s.io/api/extensions/v1beta1"
+)
+
+// predicate abstracts conditional logic w.r.t the deployment instance
+//
+// NOTE:
+// predicate is a functional approach versus traditional approach to mix
+// conditions such as *if-else* within blocks of business logic
+//
+// NOTE:
+// predicate approach enables clear separation of conditionals from
+// imperatives i.e. actions that form the business logic
+type predicate func(*deploy) bool
+
+// deployBuildOption is a typed function that abstracts anykind of operation
+// against the provided deployment instance
+//
+// This is the basic building block to create functional operations
+// against the deployment instance
+type deployBuildOption func(*deploy)
+
+// deploy is a wrapper over extnv1beta1.Deployment
+type deploy struct {
+	object *extnv1beta1.Deployment // kubernetes deployment instance
+	checks []predicate             // predicate list for deploy
+}
+
+// predicateName type is wrapper over string.
+// It is used to refer predicate and status msg.
+type predicateName string
+
+const (
+	// PredicateProgressDeadlineExceeded refer to
+	// predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded predicateName = "ProgressDeadlineExceeded"
+	// PredicateNotSpecSynced refer to predicate IsNotSpecSynced
+	PredicateNotSpecSynced predicateName = "NotSpecSynced"
+	// PredicateOlderReplicaActive refer to predicate IsOlderReplicaActive
+	PredicateOlderReplicaActive predicateName = "OlderReplicaActive"
+	// PredicateTerminationInProgress refer to predicate IsTerminationInProgress
+	PredicateTerminationInProgress predicateName = "TerminationInProgress"
+	// PredicateUpdateInProgress refer to predicate IsUpdateInProgress.
+	PredicateUpdateInProgress predicateName = "UpdateInProgress"
+)
+
+// New returns a new instance of deploy meant for deployment
+func New(opts ...deployBuildOption) *deploy {
+	k := &deploy{}
+	for _, o := range opts {
+		o(k)
+	}
+	return k
+}
+
+// WithAPIObject is a deployBuildOption caller can pass deployment schema
+// with this function to create deploy object
+func WithAPIObject(deployment *extnv1beta1.Deployment) deployBuildOption {
+	return func(d *deploy) {
+		d.object = deployment
+	}
+}
+
+// AddCheck adds the predicate as a condition to be validated
+// against the deployment instance
+func (d *deploy) AddCheck(p predicate) *deploy {
+	d.checks = append(d.checks, p)
+	return d
+}
+
+// AddChecks adds the provided predicates as conditions to be
+// validated against the deployment instance
+func (d *deploy) AddChecks(p []predicate) *deploy {
+	for _, check := range p {
+		d.AddCheck(check)
+	}
+	return d
+}
+
+// isRollout range over rolloutChecks map and check status of each predicate
+// also it generates status message from rolloutStatuses using predicate key
+func (d *deploy) isRollout() (string, bool) {
```

Comment:
IMO by trying to reuse `ok` we are making this logic difficult to understand.

Below is a sample logic that I would prefer:
```go
func (d *deploy) isRollout() (string, bool) {
	for name, predicate := range rolloutChecks {
		if predicate(d) {
			return rolloutStatuses[name](d), false
		}
	}
	return "", true
}
```

Reply:
Please check for this logic in other package as well and redesign if required

Reply:
done